Data_Engine/Query/Linq/FilterBy.cs

```diff
@@ -0,0 +1,162 @@
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+
+using BH.oM.Base;
+using BH.oM.Reflection.Attributes;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using BH.Engine.Reflection;
+
+namespace BH.Engine.Data
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Filters a list of objects to output only those with a property value matching that input")]
+        [Input("objects", "List of objects to be filtered. All objects in the list should be of similar type")]
+        [Input("propertyName", "The name of the property to filter the objects by")]
+        [Input("value", "The value of the selected property to filter the objects by")]
+        [Input("ignoreStringCase", "Ignore upper/lower case letters if the property filtered by is a text/string.")]
+        [Input("exactStringMatch", "If true checks if the property filtered by contains the text value (filtering must be a string/text)")]
+        [Output("filteredObjects", "The collection of objects that match the given filter conditions")]
+        public static List<T> FilterBy<T>(this List<T> objects, string propertyName = null, object value = null, bool ignoreStringCase = false, bool exactStringMatch = true)
+        {
+            if (objects == null || objects.Count == 0)
+            {
+                BH.Engine.Reflection.Compute.RecordWarning("No objects submitted to filter");
+                return new List<T>();
+            }
+
+            if (propertyName == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("propertyName cannot be null in order to filter the objects");
+                return new List<T>();
+            }
+
+            if (value == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("value cannot be null to filter the objects");
+                return new List<T>();
+            }
+
+            object firstObject = objects.Where(x => x != null).FirstOrDefault();
+            if(firstObject == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("All objects in the list to filter are null, please try with valid objects");
+                return new List<T>();
+            }
+
+            System.Type type = PropertyType(firstObject, propertyName);
+
+            List<T> filteredObjects = new List<T>();
+
+            if (type == typeof(System.String))
+            {
+                if (ignoreStringCase)
+                    filteredObjects.AddRange(objects.Where(x => x.PropertyValue(propertyName).ToString().Equals(value.ToString(), System.StringComparison.CurrentCultureIgnoreCase)).ToList());
```

Comment:
Assuming that you have removed the null objects already, there is still a risk that `x.PropertyValue(propertyName)` might be null